sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/RowLevelOperationRuntimeGroupFiltering.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.dynamicpruning
+
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, DynamicPruningSubquery, Expression, PredicateHelper, V2ExpressionUtils}
+import org.apache.spark.sql.catalyst.expressions.Literal.TrueLiteral
+import org.apache.spark.sql.catalyst.planning.GroupBasedRowLevelOperation
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.read.SupportsRuntimeV2Filtering
+import org.apache.spark.sql.execution.datasources.v2.{DataSourceV2Implicits, DataSourceV2Relation, DataSourceV2ScanRelation}
+
+/**
+ * A rule that assigns a subquery to filter groups in row-level operations at runtime.
+ *
+ * Data skipping during job planning for row-level operations is limited to expressions that can be
+ * converted to data source filters. Since not all expressions can be pushed down that way and
+ * rewriting groups is expensive, Spark allows data sources to filter group at runtime.
+ * If the primary scan in a group-based row-level operation supports runtime filtering, this rule
+ * will inject a subquery to find all rows that match the condition so that data sources know
+ * exactly which groups must be rewritten.
+ *
+ * Note this rule only applies to group-based row-level operations.
+ */
+case class RowLevelOperationRuntimeGroupFiltering(optimizeSubqueries: Rule[LogicalPlan])
```

Comment:
why do we need to pass the rule as a parameter? Can't we call `OptimizeSubqueries` directly in this rule?

Reply:
I also thought about this, but I think it's hard to reference `OptimizeSubqueries` outside `Optimizer`, since the former is more like a "inner class" of the latter, and references the current instance of latter in itself (i.e.: `Optimizer.this.execute(Subquery.fromExpression(s))`)

Reply:
@sunchao is correct. It wasn't easy to call `OptimizeSubqueries` outside `Optimizer`. Hence, I had to come up with this workaround.

@cloud-fan, I also considered simply adding `OptimizeSubqueries` to the batch with runtime partition filtering. However, SPARK-36444 specifically removed it from there.

Reply:
Like I said [here](https://github.com/apache/spark/pull/36304#discussion_r1014850793), we could move the new rule into a separate batch and add `OptimizeSubqueries` to it.

Reply:
An alternative idea could be to move `OptimizeSubqueries` into its own file. However, that's tricky too as it calls the optimizer.

```
Optimizer.this.execute(Subquery.fromExpression(s))
```